### PR TITLE
SIP: scuzz: add Info struct and support functions

### PR DESCRIPTION
**Pull request by rminnich**
The Info struct contains information that is drive-independent.

### CODE REVIEW

File: pkg/scuzz/sg_linux.go [hunk 1]
```diff
@@ -112,24 +112,30 @@ type packet struct {
...
+// lba48 device.
 func NewSGDisk(n string) (Disk, error) {
 	f, err := os.OpenFile(n, os.O_RDWR, 0)
 	if err != nil {
 		return nil, err
 	}
-	// TODO: figure out the dev and packID, or if we even need them.
-	return &SGDisk{f: f}, nil
+	d := &SGDisk{f: f}
+	if _, err := d.Identify(15000); err != nil {
```
Comment (hugelgupf) [inline pkg/scuzz/sg_linux.go]: What's 15,000

Comment (hugelgupf) [reply pkg/scuzz/sg_linux.go]: Oh.

Can you make timeout a property of SGDisk, since every command will have it?

Also, use time.Duration? Missed it last time.

Comment (rminnich) [reply pkg/scuzz/sg_linux.go]: Yep, yeah, Duration!

I'll make it a property of sgdisk and probably use the option pattern for NewSGDIsk?. We'll see in the long term if people need to change it a lot for different operations in the same program but ... I doubt very much they will.  That's why I left it a function parameter but ... realistically .. I bet Nobody Cares :-)

Comment (hugelgupf) [reply pkg/scuzz/sg_linux.go]: yeah, option pattern sgtm.

File: cmds/exp/hdparm/hdparm.go [hunk 2]
```diff
@@ -90,14 +95,26 @@ func main() {
...
 	scuzz.Debug = debug
 
+	timeout, err := time.ParseDuration(*timeoutDuration)
+	if err != nil {
+		log.Fatal(err)
+	}
+
 	for _, n := range flag.Args() {
-		d, err := scuzz.NewSGDisk(n)
+		d, err := scuzz.NewSGDisk(n, func(s *scuzz.SGDisk) {
```
Comment (hugelgupf) [inline cmds/exp/hdparm/hdparm.go]: can you make that a util function in scuzz? just so it reads nicer here

```go
func WithTimeout(d time.Duration) SGDiskOpt {

}
```

and

```go
scuzz.NewSGDisk(n, scuzz.WithTimeout(timeout))
```

File: pkg/dhclient/dhclient.go [hunk 3]
```diff
@@ -204,7 +204,7 @@ func lease6(ctx context.Context, iface netlink.Link, c Config) (Lease, error) {
 	}
 
 	log.Printf("Attempting to get DHCPv6 lease on %s", iface.Attrs().Name)
-	p, err := client.RapidSolicit(ctx, dhcpv6.WithNetboot)
+	p, err := client.Solicit(ctx, dhcpv6.WithNetboot)
```
Comment (hugelgupf) [inline pkg/dhclient/dhclient.go]: If you want to make this change, it's gotta be a different PR, I don't think this belongs in here

File: pkg/scuzz/ata.go [hunk 4]
```diff
@@ -78,11 +84,81 @@ type (
...
+	}{
+		{0, 0x8000, 0x8000},
+		{49, 0x200, 0x200},
+		{83, 0xc000, 0x4000},
+		{86, 0x400, 0x400},
+	}
+	for _, c := range check {
+		v, m, b := w[c.off], c.mask, c.bit
+		if (v & m) != b {
+			return fmt.Errorf("Word at offset %d: 0x%#x and should be 0x%#x", c.off, c.mask&v, b)
```
Comment (hugelgupf) [inline pkg/scuzz/ata.go]: can you make the error more descriptive for what it is? like `fmt.Errorf("not a post-2003 LBA device (your stuff too old, yo): word at offset %d ...")`

File: pkg/scuzz/ata.go [hunk 5]
```diff
@@ -78,11 +84,81 @@ type (
...
+
 	// commandDataBlock defines a command and any associated data.
 	commandDataBlock [ata16Len]byte
 
 	// statusBlock is the status returned from a drive operation.
 	statusBlock [maxStatusBlockLen]byte
 
 	direction int32
+
+	// ataString is a 10 words, bytes swapped.
```
Comment (hugelgupf) [inline pkg/scuzz/ata.go]: bytes swapped? what does that mean

File: pkg/scuzz/ata_test.go [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+import (
+	"testing"
+)
+
+func TestAtaString(t *testing.T) {
+	const want = "Copyright 2019 the u"
+	var ts ataString
+	for i := 0; i < len(ts); i++ {
+		ts[i] = uint16(want[i*2+1])<<8 | uint16(want[i*2])
```
Comment (hugelgupf) [inline pkg/scuzz/ata_test.go]: the test should do this manually -- can you just write out by hand what you expect

File: pkg/scuzz/control.go [hunk 7]
```diff
@@ -4,11 +4,43 @@
...
-	// Unlock unlocks the drive, given a password
-	Unlock(string, uint, bool) error
-	// Identify returns drive identity information
-	Identify(timeout uint) error
+import (
+	"encoding/json"
+	"fmt"
+)
+
+type (
```
Comment (hugelgupf) [inline pkg/scuzz/control.go]: super duper nit: grouped type statements always looked weird to me

File: pkg/scuzz/control.go [hunk 8]
```diff
@@ -4,11 +4,43 @@
...
+		FirmwareRevision string
+	}
+
+	// Disk is the interface to a disk, with operations
+	// to create packets and operate on them.
+	Disk interface {
+		// Unlock unlocks the drive, given a password
+		// and an indication of whether it is the
+		// master (true) or user (false) password.
+		Unlock(string, bool) error
```
Comment (hugelgupf) [inline pkg/scuzz/control.go]: can you name the arguments here, for documentation sake?

File: pkg/scuzz/control.go [hunk 9]
```diff
@@ -4,11 +4,43 @@
...
+		SectorSize       uint
+		ECCBytes         uint
+		Serial           string
+		Model            string
+		FirmwareRevision string
+	}
+
+	// Disk is the interface to a disk, with operations
+	// to create packets and operate on them.
+	Disk interface {
```
Comment (hugelgupf) [inline pkg/scuzz/control.go]: I've got a dumb question... but do you ever expect there to be an implementation of `Disk` that isn't `SGDisk`? Are there other Linux interfaces like this we'd implement it for?

File: pkg/scuzz/sg_linux.go [hunk 10]
```diff
@@ -112,24 +113,42 @@ type packet struct {
...
 }
 
 // SGDisk implements a Disk using the Linux SG device
 type SGDisk struct {
 	f      *os.File
 	dev    uint8
 	packID uint32
-	master bool
+
+	// TimeOut is the timeout on a disk operation.
```
Comment (hugelgupf) [inline pkg/scuzz/sg_linux.go]: s/TimeOut/Timeout/

File: pkg/scuzz/sg_linux.go [hunk 11]
```diff
@@ -112,24 +113,42 @@ type packet struct {
...
 	if err != nil {
 		return nil, err
 	}
-	// TODO: figure out the dev and packID, or if we even need them.
-	return &SGDisk{f: f}, nil
+	s := &SGDisk{f: f}
+	if _, err := s.Identify(); err != nil {
+		return nil, err
+	}
+	s.TimeOut, err = time.ParseDuration("15000ms")
```
Comment (hugelgupf) [inline pkg/scuzz/sg_linux.go]: `s.Timeout = 15 * time.Second`?

Comment (hugelgupf) [reply pkg/scuzz/sg_linux.go]: Also, set the timeout before Identify?

File: pkg/scuzz/sg_linux_test.go [hunk 12]
```diff
@@ -134,14 +135,18 @@ func TestUnlock(t *testing.T) {
...
-				0x00, 0x00, 0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37, 0x38, 0x39, 0x30, 0x31, 0x32, 0x33, 0x34,
+				0x00, 0x01, 0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37, 0x38, 0x39, 0x30, 0x31, 0x32, 0x33, 0x34,
 				0x35, 0x36, 0x37, 0x38, 0x39, 0x30, 0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37, 0x38, 0x39, 0x30,
 				0x31, 0x32,
 			},
 		}
 		//sb statusBlock
 	)
-	p := (&SGDisk{dev: 0x40}).unlockPacket("12345678901234567890123456789012", 15000, true)
+	timeout, err := time.ParseDuration("15000ms")
```
Comment (hugelgupf) [inline pkg/scuzz/sg_linux_test.go]: timeout := 15 * time.Second

File: pkg/scuzz/sg_linux_test.go [hunk 13]
```diff
@@ -134,14 +135,18 @@ func TestUnlock(t *testing.T) {
...
+	if err != nil {
+		t.Fatalf("time.ParseDuration(\"15000ms\"): got %v, want nil", err)
+	}
+	p := (&SGDisk{dev: 0x40, TimeOut: timeout}).unlockPacket("12345678901234567890123456789012", true)
```
Comment (hugelgupf) [inline pkg/scuzz/sg_linux_test.go]: split this into 2 lines to make more readable pls

File: pkg/scuzz/sg_linux_test.go [hunk 14]
```diff
@@ -180,6 +185,10 @@ func TestIdentify(t *testing.T) {
 		// TODO: check status block. Requires a qemu device that supports these operations.
 		//sb = statusBlock{0x70, 0x00, 0x05, 0x00, 0x00, 0x00, 0x00, 0x0a, 0x00, 0x00, 0x00, 0x00, 0x20, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}
 	)
-	p := (&SGDisk{dev: 0x40}).identifyPacket(15000)
+	timeout, err := time.ParseDuration("15000ms")
```
Comment (hugelgupf) [inline pkg/scuzz/sg_linux_test.go]: timeout := 15 * time.Second

File: pkg/dhclient/dhclient.go (hunk 3 above)
Comment (rminnich) [reply pkg/dhclient/dhclient.go]: accident :-)

File: cmds/exp/hdparm/hdparm.go (hunk 2 above)
Comment (rminnich) [reply cmds/exp/hdparm/hdparm.go]: done

File: pkg/scuzz/ata_test.go (hunk 6 above)
Comment (rminnich) [reply pkg/scuzz/ata_test.go]: I can do it but it actually is harder to read. But, ok.

File: pkg/scuzz/control.go (hunk 7 above)
Comment (rminnich) [reply pkg/scuzz/control.go]: ack.

File: pkg/scuzz/control.go (hunk 9 above)
Comment (rminnich) [reply pkg/scuzz/control.go]: yes. I am hoping and hoping that at some point sysfs gets way better, and on bsd and darwin there is no sgdisk at all. It's not one of your all time great interfaces.

File: pkg/scuzz/sg_linux.go (hunk 11 above)
Comment (rminnich) [reply pkg/scuzz/sg_linux.go]: yeah. I blanked totally one time.Second for some reason. Fixed.

File: pkg/dhclient/dhclient.go (hunk 3 above)
Comment (rminnich) [reply pkg/dhclient/dhclient.go]: THIS TIME FOR SURE .... I MEAN IT ....
